File: services/frontend-service/src/js/controllers/contribucions.controller.js
/**
 * ParkLive – contribucions.controller.js
 * Controlador de contribucions: creació de noves contribucions
 * i historial de contribucions de l'usuari.
 * Consumeix l'API Python (Flask) via AJAX.
 */

import { pythonApi } from '../api.js';
import { DEFAULT_LIMIT } from '../config.js';
import {
  showAlert,
  hideAllAlerts,
  serializeForm,
  validateForm,
  setFormLoading,
  formatDate,
  getUserId,
} from '../utils.js';

/*  FUNCIONS D'ACCÉS A L'API                                           */

/**
 * Obté l'historial de contribucions de l'usuari actual.
 *
 * @param {Object} [filtres] – Claus opcionals:
 *   tipus, validada (true/false), limit, offset
 * @returns {Promise<{total: number, contribucions: Array}>}
 */
export async function obtenirContribucionsUsuari(filtres = {}) {
  const usuariId = getUserId();
  if (!usuariId) throw new Error('Usuari no autenticat');

  const params = {
    usuari_id: usuariId,
    limit: DEFAULT_LIMIT,
    offset: 0,
    ...filtres,
  };

  try {
    const data = await pythonApi.get('/api/usuari/contribucions', params);

    return {
      total: data?.total ?? 0,
      contribucions: Array.isArray(data?.contribucions)
        ? data.contribucions
        : (Array.isArray(data) ? data : []),
    };
  } catch (err) {
    console.error('[ParkLive] Error obtenint contribucions:', err);
    throw err;
  }
}

/**
 * Crea una nova contribució (report d'estat d'un aparcament).
 *
 * @param {Object} dades – Camps obligatoris: usuari_id, aparcament_id, tipus
 *   Opcionals: estat_reportat, dades, latitud, longitud
 * @returns {Promise<Object>} – Resposta amb message i contribucio
 */
export async function crearContribucio(dades) {
  const required = ['usuari_id', 'aparcament_id', 'tipus'];
  const missing = required.filter((k) => !dades[k] && dades[k] !== 0);

  if (missing.length > 0) {
    throw new Error(`Falten camps obligatoris: ${missing.join(', ')}`);
  }

  try {
    return await pythonApi.post('/api/contribucions', dades);
  } catch (err) {
    console.error('[ParkLive] Error creant contribució:', err);
    throw err;
  }
}

/*  RENDERITZACIÓ                                                       */

/** Tipus de contribució amb icones */
const TIPUS_ICONS = {
  disponibilitat: '🅿️',
  incidencia: '⚠️',
  valoracio: '⭐',
  foto: '📷',
};

/**
 * Renderitza una llista de contribucions dins d'un contenidor.
 *
 * @param {Array}       contribucions
 * @param {HTMLElement}  container
 */
export function renderContribucions(contribucions, container) {
  if (!container) return;

  if (!contribucions || contribucions.length === 0) {
    container.innerHTML = `
      <div class="empty-state">
        <p>No s'han trobat contribucions.</p>
      </div>`;
    return;
  }

  container.innerHTML = `
    <ul class="list list-contribucions">
      ${contribucions
        .map(
          (c) => `
        <li class="list-item contribucio-item ${c.validada ? 'is-validated' : ''}">
          <span class="contribucio-icon">${TIPUS_ICONS[c.tipus] || '📝'}</span>
          <div class="contribucio-body">
            <strong class="contribucio-tipus">${escapeHtml(c.tipus || '—')}</strong>
            <span class="contribucio-estat">${escapeHtml(c.estat_reportat || '')}</span>
            ${c.dades?.comentari ? `<p class="contribucio-comentari">${escapeHtml(c.dades.comentari)}</p>` : ''}
            <small class="contribucio-meta">
              Aparcament: ${c.aparcament_id || '—'}
              · ${formatDate(c.data_creacio || c.created_at)}
              ${c.validada ? ' · ✅ Validada' : ''}
            </small>
          </div>
        </li>`,
        )
        .join('')}
    </ul>`;
}

/*  INICIALITZACIÓ DE LA PÀGINA                                        */

/**
 * Punt d'entrada del controlador de contribucions.
 * Detecta elements DOM i vincula events.
 */
export async function initContribucions() {
  const listContainer = document.querySelector('[data-role="contribucions-list"]');
  const createForm = document.querySelector('[data-role="crear-contribucio"]');
  const filterForm = document.querySelector('[data-role="contribucions-filter"]');

  // ── Llistat de contribucions ──────────────────────────────────
  if (listContainer) {
    await carregarContribucions(listContainer);
  }

  // ── Filtre ────────────────────────────────────────────────────
  if (filterForm) {
    filterForm.addEventListener('submit', async (e) => {
      e.preventDefault();
      hideAllAlerts();

      const filtres = Object.fromEntries(
        [...new FormData(filterForm).entries()].filter(([, v]) => v !== ''),
      );

      try {
        const { contribucions } = await obtenirContribucionsUsuari(filtres);
        renderContribucions(contribucions, listContainer);
      } catch (err) {
        showAlert('error', 'Error al filtrar les contribucions.');
      }
    });
  }

  // ── Formulari de crear contribució ────────────────────────────
  if (createForm) {
    // Intentar obtenir la geolocalització per omplir latitud/longitud
    prefillGeolocation(createForm);

    createForm.addEventListener('submit', async (e) => {
      e.preventDefault();
      hideAllAlerts();

      if (!validateForm(createForm)) return;

      const data = serializeForm(createForm);

      // Assegurar que l'usuari_id està present
      data.usuari_id = data.usuari_id || getUserId();

      // Convertir camps numèrics
      if (data.aparcament_id) data.aparcament_id = Number(data.aparcament_id);
      if (data.latitud) data.latitud = Number(data.latitud);
      if (data.longitud) data.longitud = Number(data.longitud);

      // Construir objecte "dades" si hi ha comentari
      if (data.comentari) {
        data.dades = { comentari: data.comentari };
        delete data.comentari;
      }

      setFormLoading(createForm, true);

      try {
        const result = await crearContribucio(data);
        showAlert('success', result.message || 'Contribució reportada amb èxit.');
        createForm.reset();

        // Refrescar llistat si existeix
        if (listContainer) {
          await carregarContribucions(listContainer);
        }
      } catch (err) {
        showAlert('error', err.message || 'Error en reportar la contribució.');
      } finally {
        setFormLoading(createForm, false);
      }
    });
  }
}

/*  HELPERS PRIVATS                                                     */

/**
 * Carrega i renderitza les contribucions de l'usuari actual.
 * @param {HTMLElement} container
 */
async function carregarContribucions(container) {
  try {
    const { contribucions } = await obtenirContribucionsUsuari();
    renderContribucions(contribucions, container);
  } catch (err) {
    showAlert('error', 'No s\'han pogut carregar les contribucions.');
  }
}

/**
 * Intenta obtenir la posició GPS del navegador i omplir
 * els camps latitud/longitud del formulari.
 *
 * @param {HTMLFormElement} form
 */
function prefillGeolocation(form) {
  if (!navigator.geolocation) return;

  const latInput = form.querySelector('[name="latitud"]');
  const lngInput = form.querySelector('[name="longitud"]');

  if (!latInput || !lngInput) return;

  navigator.geolocation.getCurrentPosition(
    (pos) => {
      latInput.value = pos.coords.latitude.toFixed(6);
      lngInput.value = pos.coords.longitude.toFixed(6);
    },
    () => {
      // L'usuari ha denegat la geolocalització – ignorar
    },
  );
}

/**
 * Escapa caràcters HTML.
 * @param {string} str
 * @returns {string}
 */
function escapeHtml(str) {
  if (!str) return '';
  const div = document.createElement('div');
  div.textContent = str;
  return div.innerHTML;
}

File: services/python-service/controllers/contribucions_controller.py
"""
Controlador per a les contribucions de disponibilitat (crowdsourcing d'usuaris).

Gestiona la creació i consulta de reports d'estat d'aparcament enviats pels
usuaris de l'aplicació, incloent validació de tipus de dades i paginació.
"""

from flask import jsonify, request
from models.contribucions_model import crear_contribucio, get_contribucions_usuari


def crear_nova_contribucio():
    """
    POST /api/contribucions - Crea una nova contribució de disponibilitat.

    Body JSON:
        usuari_id (int): ID de l'usuari reporter (obligatori).
        estat_reportat (str): 'lliure' o 'ocupat' (obligatori).
        dades (dict|None): Dades addicionals opcionals (comentari, foto_url).
        latitud (float|None): Coordenada latitud de la ubicació (opcional).
        longitud (float|None): Coordenada longitud de la ubicació (opcional).

    Returns:
        JSON 201: Contribució creada amb les dades persistides.
        JSON 400: Si falten camps obligatoris o hi ha errors de tipus.
        JSON 500: Error intern del servidor.
    """
    try:
        if not request.is_json:
            return jsonify({"error": "El contingut ha de ser JSON"}), 400

        data = request.get_json()

        required_fields = ['usuari_id', 'estat_reportat']
        missing_fields = [field for field in required_fields if field not in data]

        if missing_fields:
            return jsonify({
                "error": f"Falten els següents camps obligatoris: {', '.join(missing_fields)}"
            }), 400

        try:
            data['usuari_id'] = int(data['usuari_id'])

            if 'latitud' in data and data['latitud']:
                data['latitud'] = float(data['latitud'])

            if 'longitud' in data and data['longitud']:
                data['longitud'] = float(data['longitud'])
        except (ValueError, TypeError):
            return jsonify({"error": "Els camps numèrics tenen tipus invàlids"}), 400

        nova_contribucio = crear_contribucio(data)

        return jsonify({
            "message": "Contribució reportada amb èxit",
            "contribucio": nova_contribucio
        }), 201

    except ValueError as e:
        return jsonify({"error": str(e)}), 400
    except Exception as e:
        return jsonify({"error": f"Error en crear la contribució: {str(e)}"}), 500


def obtenir_contribucions_usuari():
    """
    GET /api/contribucions - Retorna l'historial de contribucions d'un usuari.

    Query params:
        usuari_id (int): ID de l'usuari (obligatori).
        limit (int): Nombre màxim de resultats (opcional).
        offset (int): Desplaçament per a la paginació (opcional).

    Returns:
        JSON 200: Total i llista de contribucions.
        JSON 400: Si falta usuari_id o els paràmetres no són numèrics.
        JSON 500: Error intern del servidor.
    """
    try:
        usuari_id = request.args.get('usuari_id')

        if not usuari_id:
            return jsonify({"error": "Falta el paràmetre 'usuari_id'"}), 400

        try:
            usuari_id = int(usuari_id)
        except (ValueError, TypeError):
            return jsonify({"error": "usuari_id ha de ser un nombre"}), 400

        filters = {}

        if request.args.get('limit'):
            try:
                filters['limit'] = int(request.args.get('limit'))
            except (ValueError, TypeError):
                return jsonify({"error": "limit ha de ser un nombre"}), 400

        if request.args.get('offset'):
            try:
                filters['offset'] = int(request.args.get('offset'))
            except (ValueError, TypeError):
                return jsonify({"error": "offset ha de ser un nombre"}), 400

        contribucions = get_contribucions_usuari(usuari_id, filters)

        return jsonify({
            "total": len(contribucions),
            "contribucions": contribucions
        }), 200

    except ValueError as e:
        return jsonify({"error": str(e)}), 400
    except Exception as e:
        return jsonify({"error": str(e)}), 500